From 52e6d8a2c44d10762c90ce640a0906c475e3b07b Mon Sep 17 00:00:00 2001
From: ArenM <aren@peacevolution.org>
Date: Mon, 14 Feb 2022 18:49:39 -0500
Subject: [PATCH sxmo-utils 2/5] Don't start a dbus session if one is already
 running

Systemd starts a dbus session bus automatically, and runs user services
on it. When we create our own session bus, we can't talk to those
services.

An alternative method of implementing this would be to inform systemd of
the of the bus we start, but I don't see why we would want to run extra
bus sessions.
---
 scripts/core/sxmo_winit.sh | 23 +++++++++++++++--------
 scripts/core/sxmo_xinit.sh | 25 ++++++++++++++++---------
 2 files changed, 31 insertions(+), 17 deletions(-)

diff --git a/scripts/core/sxmo_winit.sh b/scripts/core/sxmo_winit.sh
index 921d65b..d7c660f 100644
--- a/scripts/core/sxmo_winit.sh
+++ b/scripts/core/sxmo_winit.sh
@@ -2,9 +2,6 @@
 # SPDX-License-Identifier: AGPL-3.0-only
 # Copyright 2022 Sxmo Contributors
 
-# shellcheck source=scripts/core/sxmo_common.sh
-. /etc/profile.d/sxmo_init.sh
-
 envvars() {
 	export SXMO_WM=sway
 	export MOZ_ENABLE_WAYLAND=1
@@ -20,12 +17,19 @@ defaults() {
 	[ -e "$HOME"/.Xresources ] && xrdb -merge "$HOME"/.Xresources
 }
 
+with_dbus() {
+	echo "$DBUS_SESSION_BUS_ADDRESS" > "$XDG_RUNTIME_DIR"/dbus.bus
+	exec /usr/bin/sway -c "$XDG_CONFIG_HOME/sxmo/sway"
+}
+
 start() {
-	# shellcheck disable=SC2016
-	dbus-run-session sh -c '
-		echo "$DBUS_SESSION_BUS_ADDRESS" > "$XDG_RUNTIME_DIR"/dbus.bus
-		/usr/bin/sway -c "$XDG_CONFIG_HOME/sxmo/sway"
-	'
+	if [ -z "$DBUS_SESSION_BUS_ADDRESS" ]; then
+		dbus-run-session -- "$0" "with_dbus"
+	else
+		# with_dbus calls exec because dbus-run-session starts it in a
+		# new shell, but we need to keep this shell; start a subshell
+		( with_dbus )
+	fi
 }
 
 cleanup() {
@@ -35,6 +39,9 @@ cleanup() {
 }
 
 init() {
+	# shellcheck source=scripts/core/sxmo_common.sh
+	. /etc/profile.d/sxmo_init.sh
+
 	_sxmo_load_environments
 	_sxmo_prepare_dirs
 	envvars
diff --git a/scripts/core/sxmo_xinit.sh b/scripts/core/sxmo_xinit.sh
index d002fc3..7b645e1 100755
--- a/scripts/core/sxmo_xinit.sh
+++ b/scripts/core/sxmo_xinit.sh
@@ -2,9 +2,6 @@
 # SPDX-License-Identifier: AGPL-3.0-only
 # Copyright 2022 Sxmo Contributors
 
-# shellcheck source=scripts/core/sxmo_common.sh
-. /etc/profile.d/sxmo_init.sh
-
 envvars() {
 	export SXMO_WM=dwm
 	# shellcheck disable=SC2086
@@ -39,14 +36,21 @@ defaultkeyboard() {
 	fi
 }
 
+with_dbus() {
+	echo "$DBUS_SESSION_BUS_ADDRESS" > "$XDG_RUNTIME_DIR"/dbus.bus
+	# shellcheck source=configs/appcfg/xinit_template
+	. "$XDG_CONFIG_HOME"/sxmo/xinit
+	exec dwm
+}
 
 start() {
-	# shellcheck disable=SC2016
-	dbus-run-session sh -c '
-		echo "$DBUS_SESSION_BUS_ADDRESS" > "$XDG_RUNTIME_DIR"/dbus.bus
-		. "$XDG_CONFIG_HOME"/sxmo/xinit
-		dwm
-	'
+	if [ -z "$DBUS_SESSION_BUS_ADDRESS" ]; then
+		dbus-run-session -- "$0" "with_dbus"
+	else
+		# with_dbus calls exec because dbus-run-session starts it in a
+		# new shell, but we need to keep this shell; start in a subshell
+		( with_dbus )
+	fi
 }
 
 cleanup() {
@@ -56,6 +60,9 @@ cleanup() {
 }
 
 init() {
+	# shellcheck source=scripts/core/sxmo_common.sh
+	. /etc/profile.d/sxmo_init.sh
+
 	_sxmo_load_environments
 	_sxmo_prepare_dirs
 	envvars
-- 
2.36.1

